sqlbase: clarify how we reuse table names

vivekmenezes opened the pull request:
We didn't have a clear scheme on how we reuse table names.
This change discusses the different inconsistency a user
can expect when they stop using a name, or reuse a name.
DROP/TRUNCATE/RENAME commands can stop using a name.
TRUNCATE/RENAME/CREATE commands can reuse a name.

vivekmenezes committed:
sqlbase: clarify how we reuse table names

We didn't have a clear scheme on how we reuse table names.
This change discusses the different inconsistency a user
can expect when they stop using a name, or reuse a name.
DROP/TRUNCATE/RENAME commands can stop using a name.
TRUNCATE/RENAME/CREATE commands can reuse a name.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/20290)
<!-- Reviewable:end -->

vivekmenezes commented:
I'm sending you this PR so we can reach a consensus on what table name reuse looks like and the inconsistencies we will have to live with.

Once we agree, I can make more changes to the existing code to fix some bugs that I've noticed while reading the code, as well as a followup PR that allows table name reuse in a transaction.

vivekmenezes commented:
Hmmm, I actually think I can make this better by refusing leases on names that are still in use. Let me rework that idea into this PR.

vivekmenezes commented:
Updated! This is ready for review. Thanks

andreimatei commented:
So how does this here proposal mesh with the desire to allow `BEGIN; DROP foo*; CREATE foo**; COMMIT;`, which I think originally prompted kicking this nest?  Am I right to understand that, once this commits, *foo* still refers to *foo** , and it only starts referring to *foo*** later on?

---

Review status: 0 of 6 files reviewed at latest revision, 4 unresolved discussions, all commit checks successful.

---

*[pkg/sql/sqlbase/structured.proto, line 558 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20290#-L-2NrPiAxIjhayzbGwx:-L-2NrPjyFyurki_0tWE:b-8qn9at) ([raw file](https://github.com/cockroachdb/cockroach/blob/7fccb140333283c43c2287242ab5376ab8ad71d6/pkg/sql/sqlbase/structured.proto#L558)):*
> This invariant is true for inconsistencies 1 and 2, but
>   // not for 3.

Not sure what this means; how can an "invariant" not be true? Is the point you're trying to make here that we'll allow 1 and 2, but not 3?

---

*[pkg/sql/sqlbase/structured.proto, line 568 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20290#-L-2OE3_jsbr0-WkvUNb:-L-2OE3aOm0o-Tj4frG4:b-kq21d4) ([raw file](https://github.com/cockroachdb/cockroach/blob/7fccb140333283c43c2287242ab5376ab8ad71d6/pkg/sql/sqlbase/structured.proto#L568)):*
> ```protobuf
>   // name B is dropped, name C is created, and name D is reassigned. name C
>   // is viable as soon as the transaction commits. When transaction at S
>   // commits, the name B is no more, and C is viable through its new assignment.
> ```

did you mean D instead of C?
And what exactly does "viable through its new assignment" mean?

---

*[pkg/sql/sqlbase/structured.proto, line 575 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20290#-L-2OvQNV0YLMyKf8sEc:-L-2OvQNV0YLMyKf8sEd:b-jc7lrf) ([raw file](https://github.com/cockroachdb/cockroach/blob/7fccb140333283c43c2287242ab5376ab8ad71d6/pkg/sql/sqlbase/structured.proto#L575)):*
> transaction at T completing

What does "completing" mean? Committing? Or some query returning a result to its client?

---

*[pkg/sql/sqlbase/structured.proto, line 583 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20290#-L-2RkRrtusCuQt73KRK:-L-2RkRrtusCuQt73KRL:bdjvbvt) ([raw file](https://github.com/cockroachdb/cockroach/blob/7fccb140333283c43c2287242ab5376ab8ad71d6/pkg/sql/sqlbase/structured.proto#L583)):*
> partial effect of a committed transaction

That sounds scary. I think we should use a different term here. Nothing about a "transaction's effects" is partial; some operations are not single transactions, though, which I guess is what you want to talk say here. Perhaps even if it would appear otherwise because we allow renames in transactions?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20290#-:-L-30F1HOsfVDW8EiWu3:b-gjf17d)*
<!-- Sent from Reviewable.io -->

vivekmenezes committed:
sql: use DrainingNames to drain table name from DROP table

DrainingNames was only used for the RENAME case to drain old
names from the cluster. This now also handles an old name left
over from a DROP.

This change requires a migration but it's not worth it.

This change also fixes a bug where if another RENAME command
ran during the name drain phase for a transaction, the
old name  would get deleted without going through a separate
drain phase for the name. Now, the Publish() command
only deletes old names seen by the publish and also checks
that the descriptor doesn't need its version incremented.

Release note: None

vivekmenezes commented:
This PR has gone through a lot of change. I was originally planning on using it to gain consensus on what we need to do to support reassignment of table names within transaction,  but that discussion is not worth having without creating a firmer foundation.

So this PR documents the current state of the world as well as makes some improvements to the way we drain names across a cluster. It also uses the drain mechanism for the DROP case.

It also discusses an anomaly in the TRUNCATE case which is worth highlighting.

Sorry for the delayed response and I appreciate you taking the time to review this PR. Thanks!

---

Review status: 0 of 12 files reviewed at latest revision, 4 unresolved discussions.

---

*[pkg/sql/sqlbase/structured.proto, line 558 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20290#-L-2NrPiAxIjhayzbGwx:-L3EnvV9VZ-2Q0RjtnR7:byz1wj7) ([raw file](https://github.com/cockroachdb/cockroach/blob/7fccb140333283c43c2287242ab5376ab8ad71d6/pkg/sql/sqlbase/structured.proto#L558)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

> This invariant is true for inconsistencies 1 and 2, but
>   // not for 3.

Not sure what this means; how can an "invariant" not be true? Is the point you're trying to make here that we'll allow 1 and 2, but not 3?
</blockquote></details>

We're now only allowing 1 & 2

---

*[pkg/sql/sqlbase/structured.proto, line 583 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20290#-L-2RkRrtusCuQt73KRK:-L3Eo24VCsjqoSjUAjLA:b-axb2pf) ([raw file](https://github.com/cockroachdb/cockroach/blob/7fccb140333283c43c2287242ab5376ab8ad71d6/pkg/sql/sqlbase/structured.proto#L583)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

> partial effect of a committed transaction

That sounds scary. I think we should use a different term here. Nothing about a "transaction's effects" is partial; some operations are not single transactions, though, which I guess is what you want to talk say here. Perhaps even if it would appear otherwise because we allow renames in transactions?
</blockquote></details>

It's a bit less scary now but still not kosher

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20290#-:-L2fp17UzcW9NbjKYzZv:bvr37b7)*
<!-- Sent from Reviewable.io -->

vivekmenezes commented:
@andreimatei I'd love to get this in. It's a lot less controversial because it discusses the present state and proposes a few changes to match that.

andreimatei commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 0 of 12 files reviewed at latest revision, 21 unresolved discussions, all commit checks successful.

---

*[pkg/sql/drop_table.go, line 322 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/20290#-L3iYAuRr-6UkR2keAnb:-L3iYAuRr-6UkR2keAnc:bgoapmw) ([raw file](https://github.com/cockroachdb/cockroach/blob/b118915b23683bccaa22f4c391589e0be4544b54/pkg/sql/drop_table.go#L322)):*
> ```Go
> 
> func (p *planner) initiateDropTable(
> 	ctx context.Context, tableDesc *sqlbase.TableDescriptor, drainName bool,
> ```

consider making the new arg an enum

---

*[pkg/sql/drop_table.go, line 322 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/20290#-L3iYnYkB6nMcNGc0sBf:-L3iYnYkB6nMcNGc0sBg:b-pyjomo) ([raw file](https://github.com/cockroachdb/cockroach/blob/b118915b23683bccaa22f4c391589e0be4544b54/pkg/sql/drop_table.go#L322)):*
> ```Go
> 
> func (p *planner) initiateDropTable(
> 	ctx context.Context, tableDesc *sqlbase.TableDescriptor, drainName bool,
> ```

new arg needs a comment. Also why would it ever be false?

---

*[pkg/sql/schema_changer.go, line 300 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/20290#-L3i_Njvln0XfTqR-AON:-L3i_Njvln0XfTqR-AOO:b4ejbun) ([raw file](https://github.com/cockroachdb/cockroach/blob/f57548d7a63b36e4b05464bb25c75655168b585a/pkg/sql/schema_changer.go#L300)):*
> ```Go
> 		// this returns, nobody has any leases on the table, nor can get new leases,
> 		// so the table will no longer be modified.
> 		if err := sc.waitToUpdateLeases(ctx, sc.tableID); err != nil {
> ```

is it safe to remove the wait given that `DropTableDesc`, called below, doesn't wait for anything?

---

*[pkg/sql/schema_changer.go, line 370 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/20290#-L3iS0cbLBMqVNzHIGDk:-L3iS0cbLBMqVNzHIGDl:bdcfmz6) ([raw file](https://github.com/cockroachdb/cockroach/blob/f57548d7a63b36e4b05464bb25c75655168b585a/pkg/sql/schema_changer.go#L370)):*
> ```Go
> 			},
> 			// Reclaim all the old names.
> 			func(txn *client.Txn) error {
> ```

you're using the `logEvent` parameter of `Publish()`, whose name is now quite inappropriate. Also the param is not documented; in this case you're relying on the order in which the two closures are called.

Generally I find the `Publish` interface funky. Honestly, rather than using it, it think the old code was better (assuming it was correct but the commit message has no motivation for the change. You can argue for it regardless of the Publish interface - deleting those name entries has to happen after some waiting, but it doesn't nee to be transactional with the updating of the descriptor. One problem I think it had was that it didn't swallow errors from the name deletes if the records were missing cause they've already been deleted?).
What I would do is try to change the `Publish()` interface - make it take a single closure that returns a descriptor and a batch.
Your call tho.

---

*[pkg/sql/truncate.go, line 123 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/20290#-L3ib_SkYvhyAXQZRIfl:-L3ib_SkYvhyAXQZRIfm:b-fhjr03) ([raw file](https://github.com/cockroachdb/cockroach/blob/b118915b23683bccaa22f4c391589e0be4544b54/pkg/sql/truncate.go#L123)):*
> ```Go
> }
> 
> // truncateTable truncates the data of a table in a single transaction. It
> ```

The more I think about how we handle truncate, the more indefensible I find it. What gives us the liberty to make this operation not be serializable? Both us and Postgres documents it as behaving as an unqualified delete (in Postgres' case without maintaining history); we are lying.
Isn't this bad? Or, if we like this, shouldn't we implement unqualified deletes the same way?

I think we can implement it as a schema change only if we make schema changes (optionally) transactional. And I think we can make schema changes transactional - you just have to restart the txn at a timestamp where nobody has a conflicting lease, and you gotta make suckers wanting to get a new lease block on you.
I think we should backtrack on this implementation, or alternatively change our documentation for it and explain the consistency implications - but good luck with finding a wording for the documentation that doesn't sound insane :)

---

*[pkg/sql/truncate.go, line 138 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/20290#-L3ifAWGonPdJwmPatK0:-L3ifAWGonPdJwmPatK1:b-z5ilh3) ([raw file](https://github.com/cockroachdb/cockroach/blob/b118915b23683bccaa22f4c391589e0be4544b54/pkg/sql/truncate.go#L138)):*
> ```Go
> 
> 	// Remove old name -> id map.
> 	// TODO(vivek): This is done without going through the draining process
> ```

this comment is way too innocent-sounding. The consequences - that truncation does not meet our usual consistency levels, should be explicit

---

*[pkg/sql/truncate.go, line 139 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/20290#-L3iYUya_RoJMViZE6E-:-L3iYUya_RoJMViZE6E0:b1n3k7x) ([raw file](https://github.com/cockroachdb/cockroach/blob/b118915b23683bccaa22f4c391589e0be4544b54/pkg/sql/truncate.go#L139)):*
> ```Go
> 	// Remove old name -> id map.
> 	// TODO(vivek): This is done without going through the draining process
> 	// which is a violation of the invariant that a name can only name
> ```

s/a name can name/a name resolves to

---

*[pkg/sql/sqlbase/structured.proto, line 575 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20290#-L-2OvQNV0YLMyKf8sEc:-L3iGUd5_I4NYuUZ_4kC:bwrfgyr) ([raw file](https://github.com/cockroachdb/cockroach/blob/7fccb140333283c43c2287242ab5376ab8ad71d6/pkg/sql/sqlbase/structured.proto#L575)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

> transaction at T completing

What does "completing" mean? Committing? Or some query returning a result to its client?
</blockquote></details>

still not clear

---

*[pkg/sql/sqlbase/structured.proto, line 570 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20290#-L3iIEiA5CaXoOJ0wjZ1:-L3iIEiA5CaXoOJ0wjZ2:b8e1nxo) ([raw file](https://github.com/cockroachdb/cockroach/blob/12ef311924a202dd96bd0815c0d76daca45532d6/pkg/sql/sqlbase/structured.proto#L570)):*
> ```protobuf
>   // making reassigning a name complicated. In particular, since a
>   // name cannot be withdrawn across a cluster in a transaction at
>   // timestamp T, three kinds of inconsistencies can occur:
> ```

s /three kinds of inconsistencies can occur/we have to worry about the following:

---

*[pkg/sql/sqlbase/structured.proto, line 576 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20290#-L3iPav5n47U0WRzgDbU:-L3iPav5n47U0WRzgDbV:b-34jc5d) ([raw file](https://github.com/cockroachdb/cockroach/blob/12ef311924a202dd96bd0815c0d76daca45532d6/pkg/sql/sqlbase/structured.proto#L576)):*
> ```protobuf
>   // 2. A table is renamed from foo to bar at T, and both names foo and bar
>   // can be used by transactions at timestamps >= T.
>   // 3. A name foo is reassigned from one table to another at T, and the name
> ```

a thing missing from this comment is the explanation about how 3 is prevented exactly.
What happens if one does the following?
```
BEGIN;
RENAME foo TO bar;
CREATE foo;
```

I assume the CREATE will fail; it'd be useful to state that clearly. As I'm also saying elsewhere, it'd also be useful to state more clearly why the CREATE always fails if it's executed before S. This then leads to the obvious question about how is a user supposed to wait for S to occur exactly. The answer, I think, is explaining why this will work (without the BEGIN from above):
```
RENAME foo TO bar;
CREATE foo;
```

---

*[pkg/sql/sqlbase/structured.proto, line 579 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20290#-L3iJIzUTbks_0bIRmgQ:-L3iJIzVFjU6ifqJuT0D:bp6fn9b) ([raw file](https://github.com/cockroachdb/cockroach/blob/12ef311924a202dd96bd0815c0d76daca45532d6/pkg/sql/sqlbase/structured.proto#L579)):*
> ```protobuf
>   // foo can reference two different tables at timestamps >= T.
>   //
>   // To support some of these inconsistencies as naturally as possible, we
> ```

what does it mean to "support an inconsistency"? Also, what does "naturally" mean?
It's not that we like these inconsistencies and we're doing work to "support" them; we're describing the behavior of a system the permits them. I'd strike this sentence.

---

*[pkg/sql/sqlbase/structured.proto, line 580 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20290#-L3iJPrm8OukD1N0jjPL:-L3iJPrm8OukD1N0jjPM:b6n0mrr) ([raw file](https://github.com/cockroachdb/cockroach/blob/12ef311924a202dd96bd0815c0d76daca45532d6/pkg/sql/sqlbase/structured.proto#L580)):*
> ```protobuf
>   //
>   // To support some of these inconsistencies as naturally as possible, we
>   // introduce a new invariant: a name can name only a single descriptor
> ```

"we introduce a new invariant" -> new wrt what? 
I think such a phrasing doesn't make sense in a code comment.
Say "The system ensures that transactions at a given timestamp resolve a name to a single descriptor. So, we'll prevent inconsistency 3) from happening, while permitting 1 and 2.".

---

*[pkg/sql/sqlbase/structured.proto, line 588 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20290#-L3iFjxiXzKnqNrkJzZu:-L3iFjxiXzKnqNrkJzZv:b-ajiqt) ([raw file](https://github.com/cockroachdb/cockroach/blob/12ef311924a202dd96bd0815c0d76daca45532d6/pkg/sql/sqlbase/structured.proto#L588)):*
> ```protobuf
>   //
>   // The transaction at T is followed by a time period when names no longer
>   // in use are drained from the system. Once the old name is drained from
> ```

what does "names no longer in use are drained from the system"? Do you mean names no longer part of the namespace are drained?

---

*[pkg/sql/sqlbase/structured.proto, line 591 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20290#-L3iLFFK_wEqxZgEAMCe:-L3iLFFK_wEqxZgEAMCf:b-w5d4j7) ([raw file](https://github.com/cockroachdb/cockroach/blob/12ef311924a202dd96bd0815c0d76daca45532d6/pkg/sql/sqlbase/structured.proto#L591)):*
> ```protobuf
>   // the system another transaction at timestamp S is executed to release
>   // the name for future use. The interval from T to S is called the name
>   // drain interval and it allows a name to be drained across the cluster
> ```

I'd strike "and it allows a name to be drained across the cluster before it can be used by another table" - how can "an interval" allow anything? I'd say something like:
""'
 The interval from T to S is called the name drain interval: If the T transaction is removing the name foo then, at timestamps above S, foo can no longer be resolved.

---

*[pkg/sql/sqlbase/structured.proto, line 596 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20290#-L3iGAQ568YKoPcPeXZW:-L3iGAQ568YKoPcPeXZX:b-qbez1w) ([raw file](https://github.com/cockroachdb/cockroach/blob/12ef311924a202dd96bd0815c0d76daca45532d6/pkg/sql/sqlbase/structured.proto#L596)):*
> Reassignment of name D is just not permitted.

Sorry, I don't understand this. You just said that D is reassigned, and then immediately say that's not permitted. What gives? I think you want to strike "and name D is reassigned", and separately talk about how reassignments are only permitted after S.

---

*[pkg/sql/sqlbase/structured.proto, line 612 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20290#-L3iNWKouJJTmetmpYbR:-L3iNWKouJJTmetmpYbS:b-qcexwz) ([raw file](https://github.com/cockroachdb/cockroach/blob/12ef311924a202dd96bd0815c0d76daca45532d6/pkg/sql/sqlbase/structured.proto#L612)):*
> ```protobuf
>   // assignment for C, and the old assignments for B.
>   //
>   // In order for transaction at S to act properly each name maintains
> ```

this block is quite confusing. Would you mind making this more concrete by listing the steps involved in a rename, showing what S and T are, and showing the state of the namespace table at various stages?

---

*[pkg/sql/sqlbase/structured.proto, line 618 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20290#-L3iGmE8oaVSOAdgtbQz:-L3iGmE8oaVSOAdgtbR-:b5m1cgd) ([raw file](https://github.com/cockroachdb/cockroach/blob/12ef311924a202dd96bd0815c0d76daca45532d6/pkg/sql/sqlbase/structured.proto#L618)):*
> ```protobuf
>   // descriptor last owning the name.
>   //
>   // Specifically, the system.namespace table entry for a name will
> ```

Or what's this future tense ("will reference") about? What stage of some schema changes are we talking about?
As above, I'd replace this with a list of the various steps and the namespace state at each step.

---

*[pkg/sql/sqlbase/structured.proto, line 621 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20290#-L3iNJ-QKSWlBtZwxZod:-L3iNJ-QKSWlBtZwxZoe:b-fgv7b2) ([raw file](https://github.com/cockroachdb/cockroach/blob/12ef311924a202dd96bd0815c0d76daca45532d6/pkg/sql/sqlbase/structured.proto#L621)):*
> ```protobuf
>   // reference the descriptor holding a prior owner of the name requiring
>   // draining. There can be a period of time during the drain interval when
>   // the database is unable to use a name that was dropped. The user
> ```

s/database/system

---

*[pkg/sql/sqlbase/structured.proto, line 622 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20290#-L3iOr5agxwFBW6klbIT:-L3iOr5agxwFBW6klbIU:bw9thz4) ([raw file](https://github.com/cockroachdb/cockroach/blob/12ef311924a202dd96bd0815c0d76daca45532d6/pkg/sql/sqlbase/structured.proto#L622)):*
> ```protobuf
>   // draining. There can be a period of time during the drain interval when
>   // the database is unable to use a name that was dropped. The user
>   // will see an error saying the name is being drained.
> ```

I don't really understand what's this error about the name being drained is about. Why do we have it? Can we get rid of it?
I think the way things should work is that, at different points between S and T, different nodes become unable to resolve the name in question. When they fail to resolve it, the error should be the same as the one that everybody gets after S - saying that the table doesn't exist. No? Why do we need a draining-specific error?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20290#-:-L3if5IryCmbfn5y7dfY:bnfp4nl)*
<!-- Sent from Reviewable.io -->

vivekmenezes commented:
Review status: 0 of 12 files reviewed at latest revision, 21 unresolved discussions.

---

*[pkg/sql/drop_table.go, line 322 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/20290#-L3iYAuRr-6UkR2keAnb:-L46k9u6z3vF9iDlZUfx:bgdtrwl) ([raw file](https://github.com/cockroachdb/cockroach/blob/b118915b23683bccaa22f4c391589e0be4544b54/pkg/sql/drop_table.go#L322)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

consider making the new arg an enum
</blockquote></details>

I'm gonna leave it as is. I've not decided yet how we're going to tackle the question of reusing a name within a transaction case and I suspect this is all going to change with that.

---

*[pkg/sql/drop_table.go, line 322 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/20290#-L3iYnYkB6nMcNGc0sBf:-L47LW5B8tIUG1cIPhdH:b-l4396t) ([raw file](https://github.com/cockroachdb/cockroach/blob/b118915b23683bccaa22f4c391589e0be4544b54/pkg/sql/drop_table.go#L322)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

new arg needs a comment. Also why would it ever be false?
</blockquote></details>

Added a comment

---

*[pkg/sql/schema_changer.go, line 300 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/20290#-L3i_Njvln0XfTqR-AON:-L47A70X3sf0JX2yhaCn:bqi79j8) ([raw file](https://github.com/cockroachdb/cockroach/blob/f57548d7a63b36e4b05464bb25c75655168b585a/pkg/sql/schema_changer.go#L300)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

is it safe to remove the wait given that `DropTableDesc`, called below, doesn't wait for anything?
</blockquote></details>

This code is called after the call to drainNames() which runs the wait as a part of Publish()

---

*[pkg/sql/schema_changer.go, line 370 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/20290#-L3iS0cbLBMqVNzHIGDk:-L46oNiHGZJaiLFTqjEG:b-ge6vm) ([raw file](https://github.com/cockroachdb/cockroach/blob/f57548d7a63b36e4b05464bb25c75655168b585a/pkg/sql/schema_changer.go#L370)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

you're using the `logEvent` parameter of `Publish()`, whose name is now quite inappropriate. Also the param is not documented; in this case you're relying on the order in which the two closures are called.

Generally I find the `Publish` interface funky. Honestly, rather than using it, it think the old code was better (assuming it was correct but the commit message has no motivation for the change. You can argue for it regardless of the Publish interface - deleting those name entries has to happen after some waiting, but it doesn't nee to be transactional with the updating of the descriptor. One problem I think it had was that it didn't swallow errors from the name deletes if the records were missing cause they've already been deleted?).
What I would do is try to change the `Publish()` interface - make it take a single closure that returns a descriptor and a batch.
Your call tho.
</blockquote></details>

I've needed to use a Publish() here to simplify the process of draining and deleting the names. With Publish() we can guarantee that the names being deleted are indeed the ones we intended to drain. I've also added a check for UpVersion which was missing in this function and have updated the comment in the commit message.

---

*[pkg/sql/truncate.go, line 123 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/20290#-L3ib_SkYvhyAXQZRIfl:-L47BDaAef7ZrNN7_55J:b-92dt2e) ([raw file](https://github.com/cockroachdb/cockroach/blob/b118915b23683bccaa22f4c391589e0be4544b54/pkg/sql/truncate.go#L123)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

The more I think about how we handle truncate, the more indefensible I find it. What gives us the liberty to make this operation not be serializable? Both us and Postgres documents it as behaving as an unqualified delete (in Postgres' case without maintaining history); we are lying.
Isn't this bad? Or, if we like this, shouldn't we implement unqualified deletes the same way?

I think we can implement it as a schema change only if we make schema changes (optionally) transactional. And I think we can make schema changes transactional - you just have to restart the txn at a timestamp where nobody has a conflicting lease, and you gotta make suckers wanting to get a new lease block on you.
I think we should backtrack on this implementation, or alternatively change our documentation for it and explain the consistency implications - but good luck with finding a wording for the documentation that doesn't sound insane :)
</blockquote></details>

Well I don't think it's as bad as it sounds. The reality is that while it is inconsistent the TRUNCATE operation is deleting all the data! so users don't expect much besides that writes that happen "after" the truncate operation are seen. I think it's okay to leave it as is and address this with whatever scheme we come up for reassigning names within transactions.

---

*[pkg/sql/truncate.go, line 138 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/20290#-L3ifAWGonPdJwmPatK0:-L47DhaizEhhJx88Qk4f:b-xm1t2i) ([raw file](https://github.com/cockroachdb/cockroach/blob/b118915b23683bccaa22f4c391589e0be4544b54/pkg/sql/truncate.go#L138)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

this comment is way too innocent-sounding. The consequences - that truncation does not meet our usual consistency levels, should be explicit
</blockquote></details>

Called this out more explicitly

---

*[pkg/sql/truncate.go, line 139 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/20290#-L3iYUya_RoJMViZE6E-:-L47AiWEqTvQxGUyaJ5S:bb6un1j) ([raw file](https://github.com/cockroachdb/cockroach/blob/b118915b23683bccaa22f4c391589e0be4544b54/pkg/sql/truncate.go#L139)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

s/a name can name/a name resolves to
</blockquote></details>

Done

---

*[pkg/sql/sqlbase/structured.proto, line 568 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20290#-L-2OE3_jsbr0-WkvUNb:-L45vgaf9R7BvLvxQ3gq:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/7fccb140333283c43c2287242ab5376ab8ad71d6/pkg/sql/sqlbase/structured.proto#L568)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

did you mean D instead of C?
And what exactly does "viable through its new assignment" mean?
</blockquote></details>

Done.

---

*[pkg/sql/sqlbase/structured.proto, line 575 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20290#-L-2OvQNV0YLMyKf8sEc:-L45qAmQz-SWBQyAyyzA:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/7fccb140333283c43c2287242ab5376ab8ad71d6/pkg/sql/sqlbase/structured.proto#L575)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

still not clear
</blockquote></details>

Done.

---

*[pkg/sql/sqlbase/structured.proto, line 570 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20290#-L3iIEiA5CaXoOJ0wjZ1:-L45vuHe8M0nGtZ63H5B:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/12ef311924a202dd96bd0815c0d76daca45532d6/pkg/sql/sqlbase/structured.proto#L570)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

s /three kinds of inconsistencies can occur/we have to worry about the following:
</blockquote></details>

Done.

---

*[pkg/sql/sqlbase/structured.proto, line 576 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20290#-L3iPav5n47U0WRzgDbU:-L45vOxUL9Iu-wGYXTqL:bji85d) ([raw file](https://github.com/cockroachdb/cockroach/blob/12ef311924a202dd96bd0815c0d76daca45532d6/pkg/sql/sqlbase/structured.proto#L576)):*
> RENAME foo TO bar;
> CREATE foo;
I've added a couple of diagrams to explain the state of the namespace as well as why these above CREATEs fail/succeed

---

*[pkg/sql/sqlbase/structured.proto, line 579 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20290#-L3iJIzUTbks_0bIRmgQ:-L45wV0TJEOq5F6Xa0mC:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/12ef311924a202dd96bd0815c0d76daca45532d6/pkg/sql/sqlbase/structured.proto#L579)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

what does it mean to "support an inconsistency"? Also, what does "naturally" mean?
It's not that we like these inconsistencies and we're doing work to "support" them; we're describing the behavior of a system the permits them. I'd strike this sentence.
</blockquote></details>

Done.

---

*[pkg/sql/sqlbase/structured.proto, line 580 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20290#-L3iJPrm8OukD1N0jjPL:-L45xB81DKEhOwzLGXKu:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/12ef311924a202dd96bd0815c0d76daca45532d6/pkg/sql/sqlbase/structured.proto#L580)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

"we introduce a new invariant" -> new wrt what? 
I think such a phrasing doesn't make sense in a code comment.
Say "The system ensures that transactions at a given timestamp resolve a name to a single descriptor. So, we'll prevent inconsistency 3) from happening, while permitting 1 and 2.".
</blockquote></details>

Done.

---

*[pkg/sql/sqlbase/structured.proto, line 588 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20290#-L3iFjxiXzKnqNrkJzZu:-L45qbx-gxcRXg8hIzhU:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/12ef311924a202dd96bd0815c0d76daca45532d6/pkg/sql/sqlbase/structured.proto#L588)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

what does "names no longer in use are drained from the system"? Do you mean names no longer part of the namespace are drained?
</blockquote></details>

Done.

---

*[pkg/sql/sqlbase/structured.proto, line 591 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20290#-L3iLFFK_wEqxZgEAMCe:-L45xZhOA-3WXb20CTl5:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/12ef311924a202dd96bd0815c0d76daca45532d6/pkg/sql/sqlbase/structured.proto#L591)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

I'd strike "and it allows a name to be drained across the cluster before it can be used by another table" - how can "an interval" allow anything? I'd say something like:
""'
 The interval from T to S is called the name drain interval: If the T transaction is removing the name foo then, at timestamps above S, foo can no longer be resolved.

</blockquote></details>

Done.

---

*[pkg/sql/sqlbase/structured.proto, line 596 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20290#-L3iGAQ568YKoPcPeXZW:-L45r9d0hpANWezBF6fb:bcwadtc) ([raw file](https://github.com/cockroachdb/cockroach/blob/12ef311924a202dd96bd0815c0d76daca45532d6/pkg/sql/sqlbase/structured.proto#L596)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

> Reassignment of name D is just not permitted.

Sorry, I don't understand this. You just said that D is reassigned, and then immediately say that's not permitted. What gives? I think you want to strike "and name D is reassigned", and separately talk about how reassignments are only permitted after S.
</blockquote></details>

I've taken out the line about D since it's confusing

---

*[pkg/sql/sqlbase/structured.proto, line 612 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20290#-L3iNWKouJJTmetmpYbR:-L45xmaahwWqxmHKWkNQ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/12ef311924a202dd96bd0815c0d76daca45532d6/pkg/sql/sqlbase/structured.proto#L612)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

this block is quite confusing. Would you mind making this more concrete by listing the steps involved in a rename, showing what S and T are, and showing the state of the namespace table at various stages?
</blockquote></details>

Done.

---

*[pkg/sql/sqlbase/structured.proto, line 618 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20290#-L3iGmE8oaVSOAdgtbQz:-L45vjxeCgEI0Nk2lch0:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/12ef311924a202dd96bd0815c0d76daca45532d6/pkg/sql/sqlbase/structured.proto#L618)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

Or what's this future tense ("will reference") about? What stage of some schema changes are we talking about?
As above, I'd replace this with a list of the various steps and the namespace state at each step.
</blockquote></details>

Done.

---

*[pkg/sql/sqlbase/structured.proto, line 621 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20290#-L3iNJ-QKSWlBtZwxZod:-L45xiVGu74O22Ht10Ty:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/12ef311924a202dd96bd0815c0d76daca45532d6/pkg/sql/sqlbase/structured.proto#L621)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

s/database/system
</blockquote></details>

Done.

---

*[pkg/sql/sqlbase/structured.proto, line 622 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20290#-L3iOr5agxwFBW6klbIT:-L45xwYESVrln8GRmrUq:bpj7y2v) ([raw file](https://github.com/cockroachdb/cockroach/blob/12ef311924a202dd96bd0815c0d76daca45532d6/pkg/sql/sqlbase/structured.proto#L622)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

I don't really understand what's this error about the name being drained is about. Why do we have it? Can we get rid of it?
I think the way things should work is that, at different points between S and T, different nodes become unable to resolve the name in question. When they fail to resolve it, the error should be the same as the one that everybody gets after S - saying that the table doesn't exist. No? Why do we need a draining-specific error?
</blockquote></details>

I'v removed the comment. This error is today returned for case 3 with the CREATE below.

```
BEGIN;
RENAME foo TO bar;
CREATE foo;
```

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20290)*
<!-- Sent from Reviewable.io -->

vivekmenezes commented:
Thanks for your very extensive comments on this PR. I really appreciate you giving it a thorough review. Thanks

---

Review status: 0 of 12 files reviewed at latest revision, 21 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20290#-:-L47M8emYgCqQ0moZe5m:b-v5m0te)*
<!-- Sent from Reviewable.io -->

vivekmenezes commented:
@andreimatei I added another test for this PR. I'd like to merge it this morning. Thanks!

andreimatei commented:
Review status: 0 of 12 files reviewed at latest revision, 6 unresolved discussions, all commit checks successful.

---

*[pkg/sql/drop_table.go, line 322 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/20290#-L3iYnYkB6nMcNGc0sBf:-L4C2-Q0adjhCdHpxloH:b-slhbi6) ([raw file](https://github.com/cockroachdb/cockroach/blob/b118915b23683bccaa22f4c391589e0be4544b54/pkg/sql/drop_table.go#L322)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary><blockquote>

Added a comment
</blockquote></details>

the comment doesn't tell me why it'd ever be false - how do some users get away without draining.

---

*[pkg/sql/schema_changer.go, line 370 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/20290#-L3iS0cbLBMqVNzHIGDk:-L4C2vbXbAwj6PCoX7y-:bmdz21g) ([raw file](https://github.com/cockroachdb/cockroach/blob/f57548d7a63b36e4b05464bb25c75655168b585a/pkg/sql/schema_changer.go#L370)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary><blockquote>

I've needed to use a Publish() here to simplify the process of draining and deleting the names. With Publish() we can guarantee that the names being deleted are indeed the ones we intended to drain. I've also added a check for UpVersion which was missing in this function and have updated the comment in the commit message.
</blockquote></details>

at the very least the `logEvent` arg name is very inappropriate. 

---

*[pkg/sql/schema_changer.go, line 333 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/20290#-L4C8xfpPVwW_IjOSUP2:-L4C8xfpPVwW_IjOSUP3:b-gd61si) ([raw file](https://github.com/cockroachdb/cockroach/blob/4a1fac373fb4085b3b9465baa495845f619c806c/pkg/sql/schema_changer.go#L333)):*
> ```Go
> 				sc.testingKnobs.OldNamesDrainedNotification()
> 			}
> 			// Check that another schema change didn't run during the
> ```

I'm sorry but I've got no idea what's going on here.
When this closure is called, we are in the middle of the draining interval, aren't we? So how could we not be "explicitly going through a drain phase for the names".
Would you mind explaining to me what kind of race you have in mind?

---

*[pkg/sql/truncate.go, line 123 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/20290#-L3ib_SkYvhyAXQZRIfl:-L4C3eZfiC-LlCUcBgsp:bwvsont) ([raw file](https://github.com/cockroachdb/cockroach/blob/b118915b23683bccaa22f4c391589e0be4544b54/pkg/sql/truncate.go#L123)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary><blockquote>

Well I don't think it's as bad as it sounds. The reality is that while it is inconsistent the TRUNCATE operation is deleting all the data! so users don't expect much besides that writes that happen "after" the truncate operation are seen. I think it's okay to leave it as is and address this with whatever scheme we come up for reassigning names within transactions.
</blockquote></details>

I don't follow.

```
BEGIN
inflight_ops = SELECT COUNT(1) FROM ops_2018_02_03 WHERE finalized = false
if inflight_ops == 0 {
  TRUNCATE ops_2018_02_03
}
END
```

Afterwards, another bro does:
```
INSERT INTO ops_2018_02_03 (id, finalized) VALUES (42, false)
```

Now the schema change runs.

Then:

```
SELECT * FROM ops_2018_02_03 WHERE id = 42 -> no row found
```

How is this not bad? :)


If TRUNCATE doesn't behave like an unqualified delete, we need to say how it behaves...

---

*[pkg/sql/sqlbase/structured.proto, line 565 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/20290#-L4C7imj1e_DmCsmXW9a:-L4C7imj1e_DmCsmXW9b:b-wovjpy) ([raw file](https://github.com/cockroachdb/cockroach/blob/aa1fc8064b65d08ea80ca081b93f97565ad64324/pkg/sql/sqlbase/structured.proto#L565)):*
> ```protobuf
>   repeated CheckConstraint checks = 20;
> 
>   // A table descriptor is named through a name map stored in the
> ```

this explanation has improved considerably from before; thanks.

One bit that I think is still missing is the mechanics of how the drain interval starts. The txn at T writes an updated descriptor without the name being drained, right? And so what's the point at which a given node stops using the name? Is it when this updated descriptor gets leased, even though the namespace table still has an entry for the old name?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20290)*
<!-- Sent from Reviewable.io -->
